core/src/server/method_response.rs

```diff
@@ -0,0 +1,571 @@
+	}
+
+	/// Create a successful borrowed response payload.
+	pub fn success_borrowed(t: &'a T) -> Self {
+		InnerResponsePayload::success_borrowed(t).into()
+	}
+
+	/// Create an error response payload.
+	pub fn error(e: impl Into<ErrorObjectOwned>) -> Self {
+		InnerResponsePayload::error(e.into()).into()
+	}
+
+	/// Create a borrowd error response payload.
+	pub fn error_borrowed(e: impl Into<ErrorObject<'a>>) -> Self {
+		InnerResponsePayload::error_borrowed(e.into()).into()
+	}
+
+	/// Consumes the [`ResponsePayload`] and produces new [`ResponsePayload`] and a future
+	/// [`MethodResponseFuture`] that will be resolved once the response has been processed.
+	///
+	/// If `notify_on_<on_success, response, error>` has been called previously it will replaced
+	/// and the previous future(s) will be resolved with error.
+	///
+	/// The [`MethodResponseFuture`] will only return Ok on a successful JSON-RPC response.
+	pub fn notify_on_success(mut self) -> (Self, MethodResponseFuture) {
+		let (tx, rx) = response_channel();
+		self.on_exit = Some(NotifyKind::Success(tx));
+		(self, rx)
+	}
+
+	/// Consumes the [`ResponsePayload`] and produces new [`ResponsePayload`] and a future
+	/// [`MethodResponseFuture`] that will be resolved once the response has been processed.
+	///
+	/// If `notify_on_<on_success, response, error>` has been called previously it will replaced
+	/// and the previous future(s) will be resolved with error.
+	///
+	/// The [`MethodResponseFuture`] will return Ok once a response is processed.
+	pub fn notify_on_response(mut self) -> (Self, MethodResponseFuture) {
+		let (tx, rx) = response_channel();
+		self.on_exit = Some(NotifyKind::All(tx));
+		(self, rx)
+	}
+
+	/// Consumes the [`ResponsePayload`] and produces new [`ResponsePayload`] and a future
+	/// [`MethodResponseFuture`] that will be resolved once the response has been processed.
+	///
+	/// If `notify_on_<on_success, response, error>` has been called previously it will replaced
+	/// and the previous future(s) will be resolved with error.
+	///
+	/// The [`MethodResponseFuture`] will only return Ok on a failed JSON-RPC response.
+	pub fn notify_on_error(mut self) -> (Self, MethodResponseFuture) {
+		let (tx, rx) = response_channel();
+		self.on_exit = Some(NotifyKind::Error(tx));
+		(self, rx)
+	}
+
+	/// Convert the response payload into owned.
+	pub fn into_owned(self) -> ResponsePayload<'static, T> {
+		ResponsePayload { inner: self.inner.into_owned(), on_exit: self.on_exit }
+	}
+}
+
+impl<'a, T> From<ErrorCode> for ResponsePayload<'a, T>
+where
+	T: Clone,
+{
+	fn from(code: ErrorCode) -> Self {
+		let err: ErrorObject = code.into();
+		Self::error(err)
+	}
+}
+
+impl<'a> ResponsePayload<'a, ()> {
+	/// Similar to [`ResponsePayload::error`] but assigns `T = ()`.
+	///
+	/// This is useful when one only want to return the error and
+	/// `T` can't resolved by rustc to avoid type annonations.
+	pub fn unit_error(e: impl Into<ErrorObjectOwned>) -> Self {
```

Comment:
/cc @jsdw @lexnv any thoughts on this one?

Basically, I would want a nice wrapper to avoid doing `ResponsePayload::<()>::error(err)` but maybe unit error is weird because it kind of reads like the error is `()`

Reply:
When you first mentioned it I assumed it meant error is `()` which is what confused me :D

That said, I'm struggling to come up with a better name! I'm ok with just doing `ResponsePayload::<()>::error(err)` too I guess; if thename for any longer you'd not be saving any chars any more anyways!